ntp-daemon/src/peer.rs

```diff
@@ -1,177 +1,272 @@
+            send_timestamp,
+            recv_timestamp,
+        );
+
+        // Handle incoming may have changed poll interval based on message, respect that change
+        self.update_poll_wait(poll_wait, system_snapshot);
+
+        let system_poll = system_snapshot.poll_interval.as_duration();
+        let accept = self.peer.accept_synchronization(
+            ntp_instant,
+            self.config.frequency_tolerance,
+            self.config.distance_threshold,
+            system_poll,
+        );
+
+        if let Some(msg) = prepare_msg_for_system(self.index, self.reset_epoch, accept, result) {
+            self.msg_for_system_sender.send(msg).await.ok();
+        }
+    }
 }
 
 #[instrument(skip(clock, config, system_snapshots, reset))]
 pub async fn start_peer<A: ToSocketAddrs + std::fmt::Debug, C: 'static + NtpClock + Send>(
+    index: usize,
     addr: A,
     clock: C,
     config: SystemConfig,
-    mut system_snapshots: watch::Receiver<SystemSnapshot>,
-    mut reset: watch::Receiver<u64>,
-) -> Result<PeerChannels, std::io::Error> {
+    msg_for_system_sender: tokio::sync::mpsc::Sender<MsgForSystem>,
+    system_snapshots: Arc<tokio::sync::RwLock<SystemSnapshot>>,
+    reset: watch::Receiver<ResetEpoch>,
+) -> Result<(), std::io::Error> {
     let socket = UdpSocket::new("0.0.0.0:0", addr).await?;
     let our_id = ReferenceId::from_ip(socket.as_ref().local_addr().unwrap().ip());
     let peer_id = ReferenceId::from_ip(socket.as_ref().peer_addr().unwrap().ip());
-    let (tx, rx) = watch::channel::<MsgForSystem>(MsgForSystem::NoMeasurement);
-
-    // channel to notify that a reset has been completed by this peer
-    let (notify_reset_send, notify_reset_receive) = watch::channel::<u64>(0);
 
     tokio::spawn(async move {
         let local_clock_time = NtpInstant::now();
-        let mut peer = Peer::new(our_id, peer_id, local_clock_time);
+        let peer = Peer::new(our_id, peer_id, local_clock_time);
 
         let poll_wait = tokio::time::sleep(std::time::Duration::default());
         tokio::pin!(poll_wait);
 
-        // we don't store the real origin timestamp in the packet, because that would leak our
-        // system time to the network (and could make attacks easier). So instead there is some
-        // garbage data in the origin_timestamp field, and we need to track and pass along the
-        // actual origin timestamp ourselves.
-        let mut last_send_timestamp = None;
-
-        // Instant last poll message was sent (used for timing the wait)
-        let mut last_poll_sent = Instant::now();
+        let mut process = PeerProcess {
+            index,
+            clock,
+            config,
+            msg_for_system_sender,
+            system_snapshots,
+            reset,
+            socket,
+            peer,
+            last_send_timestamp: None,
+            last_poll_sent: Instant::now(),
+            reset_epoch: ResetEpoch::default(),
+        };
 
         loop {
```

Comment:
I'd move the `loop` here into a `run()` method on the `PeerProcess` type.

If I seem to be contradicting myself here re inlining single-use functions, I think that's warranted here because the behavior for the first part here (initialize a tokio task) is quite different from the second part (continually keep running the peer's workload).